### PR TITLE
Use keytab and principal to login kerberos

### LINKED ISSUE #19580: [Enhancement] Hive catalog kerberos authentication scenario optimize.
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/doris/issues?q=is%3Aissue) and found no similar issues.

### Description

当前kerberos场景hive catalog需要手动在be节点进行kinit登录，在生产环境使用上阻碍较大，需要优化。

### Solution

1、在hadoop内核代码中增加基于principal, keytab的c场景认证。
2、基于修改后的hadoop代码编译native库，doris编译时引用。
3、doris适配新的认证方式。

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by fsilent**
# Proposed changes

Issue Number: close https://github.com/apache/doris/issues/19580

## Problem summary

Describe your changes.
User keytab and princpal to login kerberos.
And user does not need to execute `kinit` manually anymore.

## Checklist(Required)

* [ ] Does it affect the original behavior
* [ ] Has unit tests been added
* [ ] Has document been added or modified
* [ ] Does it need to update dependencies
* [ ] Is this PR support rollback (If NO, please explain WHY)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: bin/start_be.sh [hunk 1]
```diff
@@ -308,6 +308,11 @@ export LIBHDFS_OPTS="${final_java_opt}"
 # see https://github.com/apache/doris/blob/master/docs/zh-CN/community/developer-guide/debug-tool.md#jemalloc-heap-profile
 export JEMALLOC_CONF="percpu_arena:percpu,background_thread:true,metadata_thp:auto,muzzy_decay_ms:30000,dirty_decay_ms:30000,oversize_threshold:0,lg_tcache_max:16,prof_prefix:jeprof.out"
 
+# We can use env vars and not need /etc/krb5.conf when KRB5_CONFIG value in env
+if [[ ${KRB5_CONFIG} ]]; then
```
Comment (morningman) [inline bin/start_be.sh:312]: No need.
You can just specify it in be.conf

Comment (fsilent) [reply bin/start_be.sh]: ok.  Deleted.

**Comment by fsilent** [comment]
run buildall

**Comment by morningman** [comment]
run buildall

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 35.68 seconds
 stream load tsv:          435 seconds loaded 74807831229 Bytes, about 164 MB/s
 stream load json:         24 seconds loaded 2358488459 Bytes, about 93 MB/s
 stream load orc:          60 seconds loaded 1101869774 Bytes, about 17 MB/s
 stream load parquet:          30 seconds loaded 861443392 Bytes, about 27 MB/s
 insert into select:          76.4 seconds inserted 10000000 Rows, about 130K ops/s
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230520073907_clickbench_pr_147670.html

**Comment by morningman** [comment]
Please patch this to pass the check:
Save it to `1.diff` and run: `patch -p1 < 1.diff

```
diff --git a/be/src/io/hdfs_builder.cpp b/be/src/io/hdfs_builder.cpp
index c0175af2d2..be1281441a 100644
--- a/be/src/io/hdfs_builder.cpp
+++ b/be/src/io/hdfs_builder.cpp
@@ -49,12 +49,12 @@ Status HDFSCommonBuilder::run_kinit() {
         return Status::InvalidArgument("Invalid hdfs_kerberos_principal or hdfs_kerberos_keytab");
     }
     std::string ticket_path = TICKET_CACHE_PATH + generate_uuid_string();
-    const char *krb_home = getenv("KRB_HOME");
+    const char* krb_home = getenv("KRB_HOME");
     std::string krb_home_str(krb_home ? krb_home : "");
     fmt::memory_buffer kinit_command;
     if (krb_home_str.empty()) {
-        fmt::format_to(kinit_command, "kinit -c {} -R -t {} -k {}", ticket_path, hdfs_kerberos_keytab,
-                hdfs_kerberos_principal);
+        fmt::format_to(kinit_command, "kinit -c {} -R -t {} -k {}", ticket_path,
+                       hdfs_kerberos_keytab, hdfs_kerberos_principal);
     } else {
         // Assign kerberos home in env, get kinit in kerberos home
         fmt::format_to(kinit_command, krb_home_str + "/bin/kinit -c {} -R -t {} -k {}", ticket_path,
@@ -67,10 +67,11 @@ Status HDFSCommonBuilder::run_kinit() {
     if (!rc) {
         return Status::InternalError("Kinit failed, errMsg: " + msg);
     }
-    #ifdef USE_LIBHDFS3
-        hdfsBuilderSetPrincipal(hdfs_builder, hdfs_kerberos_principal.c_str());
-    #endif
-        hdfsBuilderConfSetStr(hdfs_builder, "hadoop.security.kerberos.ticket.cache.path", ticket_path.c_str());
+#ifdef USE_LIBHDFS3
+    hdfsBuilderSetPrincipal(hdfs_builder, hdfs_kerberos_principal.c_str());
+#endif
+    hdfsBuilderConfSetStr(hdfs_builder, "hadoop.security.kerberos.ticket.cache.path",
+                          ticket_path.c_str());
     return Status::OK();
 }

@@ -116,22 +117,28 @@ Status createHDFSBuilder(const THdfsParams& hdfsParams, HDFSCommonBuilder* build
         hdfsBuilderSetUserName(builder->get(), hdfsParams.hdfs_kerberos_principal.c_str());
     } else if (hdfsParams.__isset.user) {
         hdfsBuilderSetUserName(builder->get(), hdfsParams.user.c_str());
+#ifdef USE_HADOOP_HDFS
         hdfsBuilderSetKerb5Conf(builder->get(), nullptr);
         hdfsBuilderSetKeyTabFile(builder->get(), nullptr);
+#endif
     }
     if (hdfsParams.__isset.hdfs_kerberos_keytab) {
         builder->need_kinit = true;
         builder->hdfs_kerberos_keytab = hdfsParams.hdfs_kerberos_keytab;
+#ifdef USE_HADOOP_HDFS
         hdfsBuilderSetKeyTabFile(builder->get(), hdfsParams.hdfs_kerberos_keytab.c_str());
+#endif
     }
     // set other conf
     if (hdfsParams.__isset.hdfs_conf) {
         for (const THdfsConf& conf : hdfsParams.hdfs_conf) {
             hdfsBuilderConfSetStr(builder->get(), conf.key.c_str(), conf.value.c_str());
+#ifdef USE_HADOOP_HDFS
             // Set krb5.conf, we should define java.security.krb5.conf in catalog properties
             if (strcmp(conf.key.c_str(), "java.security.krb5.conf") == 0) {
                 hdfsBuilderSetKerb5Conf(builder->get(), conf.value.c_str());
             }
+#endif
         }
     }

```

**Comment by fsilent** [comment]
run buildall

**Review by morningman — APPROVED** [review]
LGTM

**Pull request merged**